zio-flow/shared/src/main/scala/zio/flow/internal/DurableIndexedStore.scala

```diff
@@ -0,0 +1,47 @@
+package zio.flow.internal
+import org.rocksdb.ColumnFamilyHandle
+import zio.rocksdb.service
+import zio.{Chunk, IO, UIO, ZIO}
+import zio.stream.ZStream
+
+import java.io.IOException
+
+object DurableIndexedStore {
+
+  final case class IndexedStoreLive(rocksDB: service.RocksDB) extends IndexedStore {
+
+    private def incrementPosition(topic: String): IO[IOException, Long] = for {
+      pos <- position(topic)
+    } yield pos + 1
+
+    private def getNamespaces(rocksDB: service.RocksDB): IO[IOException, Map[Chunk[Byte], ColumnFamilyHandle]] =
+      rocksDB.initialHandles
+        .map(_.map(namespace => Chunk.fromArray(namespace.getName) -> namespace).toMap)
+        .refineToOrDie[IOException]
+
+    def getColFamilyHandle(namespace: String): UIO[ColumnFamilyHandle] =
+      for {
+        namespaces <- getNamespaces(rocksDB).orDie
+        cf         <- ZIO.succeed(namespaces(Chunk.fromArray(namespace.getBytes)))
+      } yield cf
+
+    override def position(topic: String): IO[IOException, Long] = for {
+      cfHandle <- getColFamilyHandle(topic)
+      position <- rocksDB.get(cfHandle, "POSITION".getBytes()).orDie
+    } yield new String(position.get).toLong // TODO: use zio-schema, protobuf encoding instead
+
+    override def put(topic: String, value: Chunk[Byte]): IO[IOException, Long] = for {
```

Comment:
I think we are going to have an issue here that as currently implemented this only supports adding to the initially defined topics, but we need to be able to add new topics on the fly (e.g. when we create a new `DurablePromise`). I believe RocksDB supports this functionality but that ZIO RocksDB does not currently support it so we will either have to add or implement directly.

Reply:
I've created a [PR](https://github.com/zio/zio-rocksdb/pull/23) to `zio-rocksdb` to add functions pertaining to `ColumnFamily`. Will use this to implement adding new topics.